In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
from matplotlib.colors import ListedColormap

In [2]:
# 출력 옵션 설정
pd.set_option('display.max_rows', 50)  # 모든 행 출력
pd.set_option('display.max_columns', 50)  # 모든 열 출력

# 옵션 원래대로 되돌리기
# pd.reset_option('display.max_rows')
# pd.reset_option('display.max_columns')

## 데이터 불러오고 모든 값이 동일한 컬럼 제거

fail이 일어난 디스크는 결측치를 선형보간으로 채워서 데이터 셋을 구축한다.

In [3]:
import pandas as pd
import numpy as np

data = pd.DataFrame()
for chunk in pd.read_csv("../ssd_data/combined_by_diskmodel/combined_MB1_preprocessed.csv", chunksize=100000):
    data = pd.concat([data, chunk], ignore_index=True)
data

# 데이터를 시간 순서대로 정렬 ('ds' 컬럼을 기준으로)
data = data.sort_values(by='ds')

data

disk_id        ds model    n_5  r_5   n_9      r_9  n_12  r_12  \
0          10005  20190101   MB1  100.0  0.0  99.0   4600.0  99.0  36.0   
25641      10071  20190101   MB1  100.0  0.0  97.0  12962.0  99.0  14.0   
25642      10131  20190101   MB1  100.0  0.0  96.0  15605.0  99.0  28.0   
25643      10135  20190101   MB1  100.0  0.0  96.0  15605.0  99.0  28.0   
25644      10149  20190101   MB1  100.0  0.0  99.0    371.0  99.0  17.0   
...          ...       ...   ...    ...  ...   ...      ...   ...   ...   
7106047    12233  20190630   MB1  100.0  0.0  99.0    131.0  99.0  10.0   
7106048     9465  20190630   MB1  100.0  0.0  96.0  16786.0  99.0  35.0   
7106049     9426  20190630   MB1  100.0  0.0  99.0   4902.0  99.0   5.0   
7106042     9882  20190630   MB1  100.0  0.0  96.0  15537.0  99.0  15.0   
7132652     9975  20190630   MB1  100.0  0.0  96.0  17397.0  99.0  22.0   

         n_177  r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  \
0         99.0    5.0  100.0  6515.0  100.0    0.0  100.0    0.0  100.0   
25641     99.0   37.0  100.0  6461.0  100.0    0.0  100.0    0.0  100.0   
25642     98.0  126.0  100.0  6499.0  100.0    0.0  100.0    0.0  100.0   
25643     98.0  124.0  100.0  6508.0  100.0    0.0  100.0    0.0  100.0   
25644     99.0    1.0  100.0  6563.0  100.0    0.0  100.0    0.0  100.0   
...        ...    ...    ...     ...    ...    ...    ...    ...    ...   
7106047  100.0    0.0  100.0  6566.0  100.0    0.0  100.0    0.0  100.0   
7106048   91.0  624.0  100.0  6532.0  100.0    0.0  100.0    0.0  100.0   
7106049   99.0   29.0  100.0  6559.0  100.0    0.0  100.0    0.0  100.0   
7106042   96.0  287.0  100.0  6536.0  100.0    0.0  100.0    0.0  100.0   
7132652   99.0   25.0  100.0  6513.0  100.0    0.0  100.0    0.0  100.0   

         r_183  n_184  r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  \
0          0.0  100.0    0.0  100.0    0.0   73.0   27.0   73.0   27.0  200.0   
25641      0.0  100.0    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0   
25642      0.0  100.0    0.0  100.0    0.0   74.0   26.0   74.0   26.0  200.0   
25643      0.0  100.0    0.0  100.0    0.0   75.0   25.0   75.0   25.0  200.0   
25644      0.0  100.0    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0   
...        ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
7106047    0.0  100.0    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0   
7106048    0.0  100.0    0.0  100.0    0.0   67.0   33.0   67.0   33.0  200.0   
7106049    0.0  100.0    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0   
7106042    0.0  100.0    0.0  100.0    0.0   73.0   27.0   73.0   27.0  200.0   
7132652    0.0  100.0    0.0  100.0    0.0   75.0   25.0   75.0   25.0  200.0   

         r_195  n_197  r_197  n_199  r_199  n_241         r_241  n_242  \
0          0.0  100.0    0.0  100.0    0.0   99.0  2.189215e+09   99.0   
25641      0.0  100.0    0.0  100.0    0.0   99.0  1.550334e+10   99.0   
25642      0.0  100.0    0.0  100.0    0.0   99.0  3.243394e+11   99.0   
25643      0.0  100.0    0.0  100.0    0.0   99.0  3.021083e+11   99.0   
25644      0.0  100.0    0.0  100.0    0.0   99.0  6.001201e+09   99.0   
...        ...    ...    ...    ...    ...    ...           ...    ...   
7106047    0.0  100.0    0.0  100.0    0.0   99.0  1.229311e+09   99.0   
7106048    0.0  100.0    0.0  100.0    0.0   99.0  1.109445e+12   99.0   
7106049    0.0  100.0    0.0  100.0    0.0   99.0  7.626667e+10   99.0   
7106042    0.0  100.0    0.0  100.0    0.0   99.0  7.242603e+11   99.0   
7132652    0.0  100.0    0.0  100.0    0.0   99.0  7.175013e+09   99.0   

                r_242  n_244  r_244  n_245    r_245  label  
0        2.041073e+09  100.0    0.0  100.0  65535.0      0  
25641    2.145444e+10  100.0    0.0  100.0  65535.0      0  
25642    2.622362e+11  100.0    0.0  100.0  65535.0      0  
25643    2.847856e+11  100.0    0.0  100.0  65535.0      0  
25644    6.013387e+09  100.0    0.0  100

In [4]:
data.shape

(7132653, 42)

In [5]:
data.columns

Index(['disk_id', 'ds', 'model', 'n_5', 'r_5', 'n_9', 'r_9', 'n_12', 'r_12',
       'n_177', 'r_177', 'n_180', 'r_180', 'n_181', 'r_181', 'n_182', 'r_182',
       'n_183', 'r_183', 'n_184', 'r_184', 'n_187', 'r_187', 'n_190', 'r_190',
       'n_194', 'r_194', 'n_195', 'r_195', 'n_197', 'r_197', 'n_199', 'r_199',
       'n_241', 'r_241', 'n_242', 'r_242', 'n_244', 'r_244', 'n_245', 'r_245',
       'label'],
      dtype='object')

단일값으로만 이루어진 행 제거

In [6]:
def remove_single_value_columns(df):
    single_value_cols = []
    for col in df.columns:
        unique_values = df[col].nunique()
        if unique_values == 1:
            single_value_cols.append(col)
    print(single_value_cols)
    df_filtered = df.drop(single_value_cols, axis=1)
    return df_filtered

In [7]:
data = remove_single_value_columns(data)

['model', 'n_197', 'r_197', 'n_244', 'r_244', 'n_245', 'r_245']


In [8]:
data

disk_id        ds    n_5  r_5   n_9      r_9  n_12  r_12  n_177  \
0          10005  20190101  100.0  0.0  99.0   4600.0  99.0  36.0   99.0   
25641      10071  20190101  100.0  0.0  97.0  12962.0  99.0  14.0   99.0   
25642      10131  20190101  100.0  0.0  96.0  15605.0  99.0  28.0   98.0   
25643      10135  20190101  100.0  0.0  96.0  15605.0  99.0  28.0   98.0   
25644      10149  20190101  100.0  0.0  99.0    371.0  99.0  17.0   99.0   
...          ...       ...    ...  ...   ...      ...   ...   ...    ...   
7106047    12233  20190630  100.0  0.0  99.0    131.0  99.0  10.0  100.0   
7106048     9465  20190630  100.0  0.0  96.0  16786.0  99.0  35.0   91.0   
7106049     9426  20190630  100.0  0.0  99.0   4902.0  99.0   5.0   99.0   
7106042     9882  20190630  100.0  0.0  96.0  15537.0  99.0  15.0   96.0   
7132652     9975  20190630  100.0  0.0  96.0  17397.0  99.0  22.0   99.0   

         r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  \
0          5.0  100.0  6515.0  100.0    0.0  100.0    0.0  100.0    0.0   
25641     37.0  100.0  6461.0  100.0    0.0  100.0    0.0  100.0    0.0   
25642    126.0  100.0  6499.0  100.0    0.0  100.0    0.0  100.0    0.0   
25643    124.0  100.0  6508.0  100.0    0.0  100.0    0.0  100.0    0.0   
25644      1.0  100.0  6563.0  100.0    0.0  100.0    0.0  100.0    0.0   
...        ...    ...     ...    ...    ...    ...    ...    ...    ...   
7106047    0.0  100.0  6566.0  100.0    0.0  100.0    0.0  100.0    0.0   
7106048  624.0  100.0  6532.0  100.0    0.0  100.0    0.0  100.0    0.0   
7106049   29.0  100.0  6559.0  100.0    0.0  100.0    0.0  100.0    0.0   
7106042  287.0  100.0  6536.0  100.0    0.0  100.0    0.0  100.0    0.0   
7132652   25.0  100.0  6513.0  100.0    0.0  100.0    0.0  100.0    0.0   

         n_184  r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  \
0        100.0    0.0  100.0    0.0   73.0   27.0   73.0   27.0  200.0    0.0   
25641    100.0    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0   
25642    100.0    0.0  100.0    0.0   74.0   26.0   74.0   26.0  200.0    0.0   
25643    100.0    0.0  100.0    0.0   75.0   25.0   75.0   25.0  200.0    0.0   
25644    100.0    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0   
...        ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
7106047  100.0    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0   
7106048  100.0    0.0  100.0    0.0   67.0   33.0   67.0   33.0  200.0    0.0   
7106049  100.0    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0   
7106042  100.0    0.0  100.0    0.0   73.0   27.0   73.0   27.0  200.0    0.0   
7132652  100.0    0.0  100.0    0.0   75.0   25.0   75.0   25.0  200.0    0.0   

         n_199  r_199  n_241         r_241  n_242         r_242  label  
0        100.0    0.0   99.0  2.189215e+09   99.0  2.041073e+09      0  
25641    100.0    0.0   99.0  1.550334e+10   99.0  2.145444e+10      0  
25642    100.0    0.0   99.0  3.243394e+11   99.0  2.622362e+11      0  
25643    100.0    0.0   99.0  3.021083e+11   99.0  2.847856e+11      0  
25644    100.0    0.0   99.0  6.001201e+09   99.0  6.013387e+09      0  
...        ...    ...    ...           ...    ...           ...    ...  
7106047  100.0    0.0   99.0  1.229311e+09   99.0  5.921141e+08      0  
7106048  100.0    0.0   99.0  1.109445e+12   99.0  1.247643e+11      0  
7106049  100.0    0.0   99.0  7.626667e+10   99.0  3.405889e+10      0  
7106042  100.0    0.0   99.0  7.242603e+11   99.0  5.255369e+11      0  
7132652  100.0    0.0   99.0  7.175013e+09   99.0  3.882591e+09      0  

[7132653 rows x 35 columns]

null값 지우고 하나의 데이터로만 이루어진 열을 제거한 데이터 프레임 shape과 columns확인

In [9]:
data.shape

(7132653, 35)

In [10]:
column = data.columns
column

Index(['disk_id', 'ds', 'n_5', 'r_5', 'n_9', 'r_9', 'n_12', 'r_12', 'n_177',
       'r_177', 'n_180', 'r_180', 'n_181', 'r_181', 'n_182', 'r_182', 'n_183',
       'r_183', 'n_184', 'r_184', 'n_187', 'r_187', 'n_190', 'r_190', 'n_194',
       'r_194', 'n_195', 'r_195', 'n_199', 'r_199', 'n_241', 'r_241', 'n_242',
       'r_242', 'label'],
      dtype='object')

In [11]:
unique_disk_ids = data['disk_id'].nunique()
print("disk_id의 고유 개수:", unique_disk_ids)

disk_id의 고유 개수: 42343


fail이 일어난 행들 확인

In [12]:
data[data['label']==1]

disk_id        ds   n_5  r_5   n_9      r_9  n_12   r_12  n_177  \
19805      32240  20190101  99.0  2.0  97.0  13752.0  99.0   20.0   99.0   
3982       17909  20190101  99.0  2.0  97.0  12819.0  99.0   19.0   97.0   
131303     44290  20190104  99.0  2.0  97.0  11289.0  99.0   16.0   97.0   
164697      8656  20190105  99.0  2.0  97.0  13352.0  99.0   16.0   98.0   
209834     12584  20190106  99.0  2.0  97.0  10339.0  99.0   19.0   92.0   
...          ...       ...   ...  ...   ...      ...   ...    ...    ...   
7075677    40280  20190629  99.0  2.0  96.0  17263.0  99.0  125.0   95.0   
7089122    37897  20190629  99.0  2.0  96.0  17436.0  99.0   18.0   95.0   
7056471    21058  20190629  99.0  4.0  97.0  14548.0  99.0   28.0   98.0   
7068033    28019  20190629  99.0  2.0  96.0  16638.0  99.0   23.0  100.0   
7097772     8599  20190630  99.0  2.0  96.0  17264.0  99.0   21.0   91.0   

         r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  \
19805     53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0   
3982     155.0   99.0  6471.0  100.0    0.0  100.0    0.0   99.0    2.0   
131303   176.0   99.0  6542.0  100.0    0.0  100.0    0.0   99.0    2.0   
164697   148.0   99.0  6483.0  100.0    0.0  100.0    0.0   99.0    2.0   
209834   564.0   99.0  6547.0  100.0    0.0  100.0    0.0   99.0    2.0   
...        ...    ...     ...    ...    ...    ...    ...    ...    ...   
7075677  308.0   99.0  6513.0  100.0    0.0  100.0    0.0   99.0    2.0   
7089122  300.0   99.0  6526.0  100.0    0.0  100.0    0.0   99.0    2.0   
7056471  136.0   99.0  6375.0  100.0    0.0  100.0    0.0   99.0    4.0   
7068033    0.0   99.0  6445.0  100.0    0.0  100.0    0.0   99.0    2.0   
7097772  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0   

         n_184  r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  \
19805    100.0    0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0   
3982     100.0    0.0   99.0    1.0   68.0   32.0   68.0   32.0  199.0    1.0   
131303   100.0    0.0  100.0    0.0   75.0   25.0   75.0   25.0  200.0    0.0   
164697   100.0    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0   
209834   100.0    0.0  100.0    0.0   74.0   26.0   74.0   26.0  200.0    0.0   
...        ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
7075677  100.0    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0   
7089122  100.0    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0   
7056471  100.0    0.0   99.0    1.0   73.0   27.0   73.0   27.0  199.0    1.0   
7068033  100.0    0.0  100.0    0.0   74.0   26.0   74.0   26.0  200.0    0.0   
7097772  100.0    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0   

         n_199  r_199  n_241         r_241  n_242         r_242  label  
19805    100.0    0.0   99.0  1.863063e+11   99.0  1.269325e+11      1  
3982     100.0    0.0   99.0  4.761304e+11   99.0  3.422788e+11      1  
131303   100.0    0.0   99.0  5.057076e+11   99.0  3.532765e+11      1  
164697   100.0    0.0   99.0  4.403823e+11   99.0  3.495673e+11      1  
209834   100.0    0.0   99.0  1.391075e+12   99.0  2.054236e+11      1  
...        ...    ...    ...           ...    ...           ...    ...  
7075677  100.0    0.0   99.0  9.733655e+11   99.0  7.904370e+11      1  
7089122  100.0    0.0   99.0  9.776624e+11   99.0  8.292570e+11      1  
7056471   99.0    2.0   99.0  4.534237e+11   99.0  2.597320e+11      1  
7068033  100.0    0.0   99.0  1.471218e+09   99.0  1.148237e+10      1  
7097772  100.0    0.0   99.0  2.201928e+12   99.0  8.381230e+11      1  

[304 rows x 35 columns]

## failed된 데이터 가져오기
이때 failed된 데이터중에서 결측치가 있거나, 없는 날짜가 있으면 보간법으로 값을 채우겠다.<br>
여기서 마지막으로 생성된 selected_filtered_data_failed는 1월1일부터 6월 30일까지 데이터가 다 존재하는 failed disk_id들만 모은것이고<br>
failed disk_id 는 304개 였지만 모든 날짜가 다 있는 disk_id는 139개이다.그리고 보간하기전에 모든 날짜가 있는 disk_id는 55개이다.

In [32]:
failed_data=data[data['label']==1]
unique_ids = failed_data['disk_id'].unique()

In [33]:
len(unique_ids)

304

In [34]:
unique_disk_ids = []  # 유일한 disk_id를 저장할 리스트

for disk_id in unique_ids:
    disk_data = data[data['disk_id'] == disk_id]
    if len(disk_data) == 180:  # 행의 개수가 180인 경우에만 추가
        unique_disk_ids.append(disk_id)

unique_disk_ids = pd.Series(unique_disk_ids, name='disk_id') # 약 3초걸림

보간하기 전에 180개의 행을 가지고 있는 disk_id를 보자

In [35]:
unique_disk_ids

0     32240
1     43191
2     28588
3     30277
4     24395
      ...  
50    27664
51    35076
52     4885
53    35074
54     6314
Name: disk_id, Length: 55, dtype: int64

fail된 disk중에서 180개가 안되는것들을 보간해보자

In [36]:
dfs_to_concat = []  # 추가할 데이터프레임들을 저장할 리스트

for disk_id in unique_ids:
    disk_data = data[data['disk_id'] == disk_id].copy()
    disk_data['ds'] = pd.to_datetime(disk_data['ds'], format='%Y%m%d')
    disk_data = disk_data.sort_values(by='ds')
    disk_data = disk_data.set_index('ds', drop=True)
    all_dates = pd.date_range(start=disk_data.index.min(), end=disk_data.index.max(), freq='D')
    disk_data = disk_data.reindex(all_dates)
    disk_data = disk_data.interpolate(method='linear')
    dfs_to_concat.append(disk_data)

filled_data_failed = pd.concat(dfs_to_concat)  # 데이터프레임들을 연결하여 새로운 데이터프레임 생성

filled_data_failed = filled_data_failed.reset_index()  # 인덱스 재설정

In [37]:
filled_data_failed

index  disk_id    n_5  r_5   n_9      r_9  n_12  r_12  n_177  \
0     2019-01-01  32240.0   99.0  2.0  97.0  13752.0  99.0  20.0   99.0   
1     2019-01-02  32240.0   99.0  2.0  97.0  13776.0  99.0  20.0   99.0   
2     2019-01-03  32240.0   99.0  2.0  97.0  13800.0  99.0  20.0   99.0   
3     2019-01-04  32240.0   99.0  2.0  97.0  13824.0  99.0  20.0   99.0   
4     2019-01-05  32240.0   99.0  2.0  97.0  13848.0  99.0  20.0   99.0   
...          ...      ...    ...  ...   ...      ...   ...   ...    ...   
43895 2019-06-26   8599.0  100.0  0.0  96.0  17168.0  99.0  21.0   91.0   
43896 2019-06-27   8599.0  100.0  0.0  96.0  17192.0  99.0  21.0   91.0   
43897 2019-06-28   8599.0  100.0  0.0  96.0  17216.0  99.0  21.0   91.0   
43898 2019-06-29   8599.0   99.0  2.0  96.0  17240.0  99.0  21.0   91.0   
43899 2019-06-30   8599.0   99.0  2.0  96.0  17264.0  99.0  21.0   91.0   

       r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  n_184  \
0       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
1       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
2       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
3       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
4       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
...      ...    ...     ...    ...    ...    ...    ...    ...    ...    ...   
43895  623.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43896  623.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43897  624.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43898  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
43899  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   

       r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  n_199  \
0        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
1        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
2        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
3        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
4        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
43895    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43896    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43897    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
43898    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43899    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   

       r_199  n_241         r_241  n_242         r_242  label  
0        0.0   99.0  1.863063e+11   99.0  1.269325e+11    1.0  
1        0.0   99.0  1.864852e+11   99.0  1.272425e+11    0.0  
2        0.0   99.0  1.866529e+11   99.0  1.275511e+11    0.0  
3        0.0   99.0  1.868283e+11   99.0  1.278591e+11    0.0  
4        0.0   99.0  1.870048e+11   99.0  1.281759e+11    0.0  
...      ...    ...           ...    ...           ...    ...  
43895    0.0   99.0  2.201000e+12   99.0  8.356453e+11    0.0  
43896    0.0   99.0  2.201312e+12   99.0  8.362658e+11    0.0  
43897    0.0   99.0  2.201585e+12   99.0  8.371074e+11    0.0  
43898    0.0   99.0  2.201871e+12   99.0  8.379656e+11    0.0  
43899    0.0   99.0  2.201928e+12   99.0  8.381230e+11    1.0  

[43900 rows x 35 columns]

1월1일부터 시작안하는 failed_disk들도 많아서  304*181 = 55024로 딱 떨어지지는 않는다.<br>
그래서 181개로 보간되는 것들을 살펴보니 138개가 있어서 failed한 데이터 138개 healthy한 데이터 138개를 가져와서 비율을 맞추겠다.

In [46]:
len(unique_ids)

304

In [48]:
unique_disk_ids = []  # 유일한 disk_id를 저장할 리스트

for disk_id in unique_ids:
    disk_data = filled_data_failed[filled_data_failed['disk_id'] == disk_id]
    if len(disk_data) == 181:  # 행의 개수가 180인 경우에만 추가
        unique_disk_ids.append(disk_id)

unique_disk_ids = pd.Series(unique_disk_ids, name='disk_id') 

In [49]:
unique_disk_ids

0      32240
1       8656
2      38576
3      43191
4      28588
       ...  
134    18461
135    40280
136    37897
137    28019
138     8599
Name: disk_id, Length: 139, dtype: int64

In [50]:
selected_filtered_data_failed = filled_data_failed[filled_data_failed['disk_id'].isin(unique_disk_ids)]
selected_filtered_data_failed

index  disk_id    n_5  r_5   n_9      r_9  n_12  r_12  n_177  \
0     2019-01-01  32240.0   99.0  2.0  97.0  13752.0  99.0  20.0   99.0   
1     2019-01-02  32240.0   99.0  2.0  97.0  13776.0  99.0  20.0   99.0   
2     2019-01-03  32240.0   99.0  2.0  97.0  13800.0  99.0  20.0   99.0   
3     2019-01-04  32240.0   99.0  2.0  97.0  13824.0  99.0  20.0   99.0   
4     2019-01-05  32240.0   99.0  2.0  97.0  13848.0  99.0  20.0   99.0   
...          ...      ...    ...  ...   ...      ...   ...   ...    ...   
43895 2019-06-26   8599.0  100.0  0.0  96.0  17168.0  99.0  21.0   91.0   
43896 2019-06-27   8599.0  100.0  0.0  96.0  17192.0  99.0  21.0   91.0   
43897 2019-06-28   8599.0  100.0  0.0  96.0  17216.0  99.0  21.0   91.0   
43898 2019-06-29   8599.0   99.0  2.0  96.0  17240.0  99.0  21.0   91.0   
43899 2019-06-30   8599.0   99.0  2.0  96.0  17264.0  99.0  21.0   91.0   

       r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  n_184  \
0       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
1       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
2       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
3       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
4       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
...      ...    ...     ...    ...    ...    ...    ...    ...    ...    ...   
43895  623.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43896  623.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43897  624.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43898  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
43899  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   

       r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  n_199  \
0        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
1        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
2        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
3        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
4        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
43895    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43896    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43897    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
43898    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43899    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   

       r_199  n_241         r_241  n_242         r_242  label  
0        0.0   99.0  1.863063e+11   99.0  1.269325e+11    1.0  
1        0.0   99.0  1.864852e+11   99.0  1.272425e+11    0.0  
2        0.0   99.0  1.866529e+11   99.0  1.275511e+11    0.0  
3        0.0   99.0  1.868283e+11   99.0  1.278591e+11    0.0  
4        0.0   99.0  1.870048e+11   99.0  1.281759e+11    0.0  
...      ...    ...           ...    ...           ...    ...  
43895    0.0   99.0  2.201000e+12   99.0  8.356453e+11    0.0  
43896    0.0   99.0  2.201312e+12   99.0  8.362658e+11    0.0  
43897    0.0   99.0  2.201585e+12   99.0  8.371074e+11    0.0  
43898    0.0   99.0  2.201871e+12   99.0  8.379656e+11    0.0  
43899    0.0   99.0  2.201928e+12   99.0  8.381230e+11    1.0  

[25159 rows x 35 columns]

## failed된 disk_id만큼 healthy한 disk_id가져오기
failed된 디스크중에서 모든 날짜가 다 존재하는 disk_id는 139개여서 healthy한 디스크 아이디 중에서 139개를 가져와서 데이터를 만들겠다.
결측치가 없는 데이터중에서 random으로 가져오겠다. 

In [ ]:
healthy_data=data[data['label']==0]
unique_ids = healthy_data['disk_id'].unique()

In [ ]:
unique_ids

array([10005, 10071, 10131, ..., 32519, 15111, 15692], dtype=int64)

In [ ]:
unique_disk_ids = []  # 유일한 disk_id를 저장할 리스트

for disk_id in unique_ids:
    disk_data = healthy_data[healthy_data['disk_id'] == disk_id]
    if len(disk_data) == 180:  # 행의 개수가 180인 경우에만 추가
        unique_disk_ids.append(disk_id)

unique_disk_ids = pd.Series(unique_disk_ids, name='disk_id') # 약 6분걸림

In [ ]:
len(unique_disk_ids)

14924

In [ ]:
unique_disk_ids

0        10005
1         1007
2         1020
3        10598
4        10682
         ...  
14919     9737
14920      987
14921     9923
14922     9545
14923     8297
Name: disk_id, Length: 14924, dtype: int64

In [ ]:
import random

random.seed(42)  # 랜덤 선택 결과 재현을 위한 시드 설정

selected_ids = random.sample(list(unique_disk_ids), 139)  # unique_disk_ids에서 랜덤으로 304개 선택

selected_disk_ids = pd.Series(selected_ids, name='disk_id')

In [ ]:
selected_disk_ids

0      29917
1       4040
2      24759
3      28302
4      12786
       ...  
134     6314
135    33543
136    28019
137     1408
138     8844
Name: disk_id, Length: 139, dtype: int64

근데 어이없게도 healthy한 데이터중에서 selected_disk_ids를 뽑았는데 이 데이터에서도 6월 13일이 없다. 애초에 모든 데이터가 6월 13일이 없다 그래서 6월 13일은 넣고 181일로 맞춰주는 작업을 하고 이때 보간법을 이용하겠다.

In [ ]:
dfs_to_concat = []  # 추가할 데이터프레임들을 저장할 리스트

for disk_id in selected_disk_ids:
    disk_data = data[data['disk_id'] == disk_id].copy()
    disk_data['ds'] = pd.to_datetime(disk_data['ds'], format='%Y%m%d')
    disk_data = disk_data.sort_values(by='ds')
    disk_data = disk_data.set_index('ds', drop=True)
    all_dates = pd.date_range(start=disk_data.index.min(), end=disk_data.index.max(), freq='D')
    disk_data = disk_data.reindex(all_dates)
    disk_data = disk_data.interpolate(method='linear')
    dfs_to_concat.append(disk_data)

selected_filtered_data_healthy = pd.concat(dfs_to_concat)  # 데이터프레임들을 연결하여 새로운 데이터프레임 생성

selected_filtered_data_healthy = selected_filtered_data_healthy.reset_index()  # 인덱스 재설정

In [ ]:
selected_filtered_data_healthy.shape

(25159, 35)

In [ ]:
selected_filtered_data_healthy

index  disk_id    n_5  r_5   n_9      r_9  n_12  r_12  n_177  \
0     2019-01-01  29917.0  100.0  0.0  97.0  14763.0  99.0  19.0   95.0   
1     2019-01-02  29917.0  100.0  0.0  97.0  14787.0  99.0  19.0   95.0   
2     2019-01-03  29917.0  100.0  0.0  97.0  14811.0  99.0  19.0   95.0   
3     2019-01-04  29917.0  100.0  0.0  97.0  14835.0  99.0  19.0   95.0   
4     2019-01-05  29917.0  100.0  0.0  97.0  14859.0  99.0  19.0   95.0   
...          ...      ...    ...  ...   ...      ...   ...   ...    ...   
25154 2019-06-26   8844.0   99.0  2.0  96.0  18993.0  99.0  18.0   95.0   
25155 2019-06-27   8844.0   99.0  2.0  96.0  19017.0  99.0  18.0   95.0   
25156 2019-06-28   8844.0   99.0  2.0  96.0  19041.0  99.0  18.0   95.0   
25157 2019-06-29   8844.0   99.0  2.0  96.0  19065.0  99.0  18.0   95.0   
25158 2019-06-30   8844.0   99.0  2.0  96.0  19089.0  99.0  18.0   95.0   

       r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  n_184  \
0      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
1      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
2      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
3      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
4      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
...      ...    ...     ...    ...    ...    ...    ...    ...    ...    ...   
25154  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25155  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25156  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25157  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25158  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   

       r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  n_199  \
0        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
1        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
2        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
3        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
4        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
25154    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25155    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25156    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25157    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25158    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   

       r_199  n_241         r_241  n_242         r_242  label  
0        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
1        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
2        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
3        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
4        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
...      ...    ...           ...    ...           ...    ...  
25154    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25155    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25156    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25157    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25158    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  

[25159 rows x 35 columns]

## 결론 
selected_filtered_data_healthy, selected_filtered_data_failed의 두개의 데이터 프레임이 있는데 각각 139개의 disk_id를 가지고 있고 <br>
139*181= 25159를 보면 알겠지만 각각 6개월에 대해 완벽한 데이터 셋이라고 볼 수 있다.

In [56]:
selected_filtered_data_healthy

index  disk_id    n_5  r_5   n_9      r_9  n_12  r_12  n_177  \
0     2019-01-01  29917.0  100.0  0.0  97.0  14763.0  99.0  19.0   95.0   
1     2019-01-02  29917.0  100.0  0.0  97.0  14787.0  99.0  19.0   95.0   
2     2019-01-03  29917.0  100.0  0.0  97.0  14811.0  99.0  19.0   95.0   
3     2019-01-04  29917.0  100.0  0.0  97.0  14835.0  99.0  19.0   95.0   
4     2019-01-05  29917.0  100.0  0.0  97.0  14859.0  99.0  19.0   95.0   
...          ...      ...    ...  ...   ...      ...   ...   ...    ...   
25154 2019-06-26   8844.0   99.0  2.0  96.0  18993.0  99.0  18.0   95.0   
25155 2019-06-27   8844.0   99.0  2.0  96.0  19017.0  99.0  18.0   95.0   
25156 2019-06-28   8844.0   99.0  2.0  96.0  19041.0  99.0  18.0   95.0   
25157 2019-06-29   8844.0   99.0  2.0  96.0  19065.0  99.0  18.0   95.0   
25158 2019-06-30   8844.0   99.0  2.0  96.0  19089.0  99.0  18.0   95.0   

       r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  n_184  \
0      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
1      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
2      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
3      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
4      318.0  100.0  6510.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
...      ...    ...     ...    ...    ...    ...    ...    ...    ...    ...   
25154  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25155  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25156  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25157  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
25158  323.0   99.0  6504.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   

       r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  n_199  \
0        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
1        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
2        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
3        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
4        0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
25154    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25155    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25156    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25157    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   
25158    0.0  100.0    0.0   72.0   28.0   72.0   28.0  200.0    0.0  100.0   

       r_199  n_241         r_241  n_242         r_242  label  
0        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
1        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
2        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
3        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
4        0.0   99.0  1.044736e+12   99.0  1.392127e+12    0.0  
...      ...    ...           ...    ...           ...    ...  
25154    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25155    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25156    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25157    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  
25158    0.0   99.0  1.031799e+12   99.0  1.375688e+12    0.0  

[25159 rows x 35 columns]

In [57]:
selected_filtered_data_failed

index  disk_id    n_5  r_5   n_9      r_9  n_12  r_12  n_177  \
0     2019-01-01  32240.0   99.0  2.0  97.0  13752.0  99.0  20.0   99.0   
1     2019-01-02  32240.0   99.0  2.0  97.0  13776.0  99.0  20.0   99.0   
2     2019-01-03  32240.0   99.0  2.0  97.0  13800.0  99.0  20.0   99.0   
3     2019-01-04  32240.0   99.0  2.0  97.0  13824.0  99.0  20.0   99.0   
4     2019-01-05  32240.0   99.0  2.0  97.0  13848.0  99.0  20.0   99.0   
...          ...      ...    ...  ...   ...      ...   ...   ...    ...   
43895 2019-06-26   8599.0  100.0  0.0  96.0  17168.0  99.0  21.0   91.0   
43896 2019-06-27   8599.0  100.0  0.0  96.0  17192.0  99.0  21.0   91.0   
43897 2019-06-28   8599.0  100.0  0.0  96.0  17216.0  99.0  21.0   91.0   
43898 2019-06-29   8599.0   99.0  2.0  96.0  17240.0  99.0  21.0   91.0   
43899 2019-06-30   8599.0   99.0  2.0  96.0  17264.0  99.0  21.0   91.0   

       r_177  n_180   r_180  n_181  r_181  n_182  r_182  n_183  r_183  n_184  \
0       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
1       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
2       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
3       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
4       53.0   99.0  6543.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
...      ...    ...     ...    ...    ...    ...    ...    ...    ...    ...   
43895  623.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43896  623.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43897  624.0  100.0  6501.0  100.0    0.0  100.0    0.0  100.0    0.0  100.0   
43898  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   
43899  624.0   99.0  6499.0  100.0    0.0  100.0    0.0   99.0    2.0  100.0   

       r_184  n_187  r_187  n_190  r_190  n_194  r_194  n_195  r_195  n_199  \
0        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
1        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
2        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
3        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
4        0.0  100.0    0.0   71.0   29.0   71.0   29.0  200.0    0.0  100.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
43895    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43896    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43897    0.0  100.0    0.0   69.0   31.0   69.0   31.0  200.0    0.0  100.0   
43898    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   
43899    0.0  100.0    0.0   70.0   30.0   70.0   30.0  200.0    0.0  100.0   

       r_199  n_241         r_241  n_242         r_242  label  
0        0.0   99.0  1.863063e+11   99.0  1.269325e+11    1.0  
1        0.0   99.0  1.864852e+11   99.0  1.272425e+11    0.0  
2        0.0   99.0  1.866529e+11   99.0  1.275511e+11    0.0  
3        0.0   99.0  1.868283e+11   99.0  1.278591e+11    0.0  
4        0.0   99.0  1.870048e+11   99.0  1.281759e+11    0.0  
...      ...    ...           ...    ...           ...    ...  
43895    0.0   99.0  2.201000e+12   99.0  8.356453e+11    0.0  
43896    0.0   99.0  2.201312e+12   99.0  8.362658e+11    0.0  
43897    0.0   99.0  2.201585e+12   99.0  8.371074e+11    0.0  
43898    0.0   99.0  2.201871e+12   99.0  8.379656e+11    0.0  
43899    0.0   99.0  2.201928e+12   99.0  8.381230e+11    1.0  

[25159 rows x 35 columns]